### PR TITLE
Context replacement

### PR DESCRIPTION
**Pull request by Polymorphe57**

### CODE REVIEW

File: src/reactivity.ts [hunk 1]
```diff
@@ -138,13 +156,54 @@ function unregisterObserver(observer: Observer) {
...
-export function useState(state: any): Atom {
+export function useState(state: any, observer?: Observer): Atom {
   if (!isTrackable(state)) {
     throw new Error("Argument is not trackable");
   }
   const node = getCurrent()!;
-  const observer = () => node.render();
-  const unregisterObserver = registerObserver(observer);
+  observer = () => node.render();
+  const unregisterObserver = registerObserver(observer, { parent: node.parent });
```
Comment (Polymorphe57) [inline src/reactivity.ts:165]: Two observers should not be associated with same node!
Alternative: share observer for same node

File: src/reactivity.ts [hunk 2]
```diff
@@ -103,22 +106,35 @@ function isTrackable(value: any): boolean {
...
   }
 }
 
-export function registerObserver(observer: Observer) {
+export function registerObserver(observer: Observer, node?: Node) {
   if (!observerAtoms.get(observer)) {
     observerAtoms.set(observer, new Set());
+    node = node || { parent: null };
+    nodeToObserver.set(node, observer);
+    add(node);
```
Comment (Polymorphe57) [inline src/reactivity.ts:137]: we create a tree with a virtual root that contains all observer nodes and their parents even if they are not observers! 
This simplifies the add operation.

File: src/reactivity.ts [hunk 3]
```diff
@@ -144,7 +160,47 @@ export function useState(state: any): Atom {
...
+
+function getMaximalObservers(observers: Set<Observer>): Set<Observer> {
+  const acc: Set<Observer> = new Set();
+  const predicate = (observer: Observer) => observers.has(observer);
+  getMaximalObserversAux(predicate, root, acc);
+  return acc;
+}
+
+function getMaximalObserversAux(predicate: (observer: Observer) => boolean, node: Node, acc: Set<Observer>) {
+  const observer = nodeToObserver.get(node);
```
Comment (Polymorphe57) [inline src/reactivity.ts:198]: not sure it is good: reminds me of some courses on tail recursive functions

File: tests/context_test.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
+import { Component, mount, onRender, onWillStart, onWillUpdateProps, useState } from "../src";
```
Comment (Polymorphe57) [inline tests/context_test.ts:1]: tests should be put in reactivity_test maybe?

File: tests/context_test.ts [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+    await nextMicroTick();
+    parent.state.flag = false;
+    await prom;
+    expect(fixture.innerHTML).toBe("<div></div>");
+    testContext.a = 456;
+    await nextMicroTick();
+    expect(steps).toEqual([]);
+  });
+
+  test.only("destroyed component before being mounted is inactive", async () => {
```
Comment (ged-odoo) [inline tests/context_test.ts:395]: hey

File: src/reactivity.ts [hunk 6]
```diff
@@ -60,19 +74,51 @@ function createAtom(source: Source, observer: Observer): Atom {
...
+function addObserverAtom(observer: Observer, atom: Atom) {
+  observerAtoms.get(observer)!.add(atom);
+}
+
+function addSourceAtom(source: Source, atom: Atom) {
+  addToNotify(sourceAtoms.get(source)!.get(ROOT)!, atom);
+}
+
+function addSourceKeyAtom(source: Source, key: any, atom: Atom) {
+  // we don't want to redo comutations here if the key was added (and the observer is registerd)
```
Comment (ged-odoo) [inline src/reactivity.ts:97]: computation

File: src/reactivity.ts [hunk 7]
```diff
@@ -143,8 +186,89 @@ export function useState(state: any): Atom {
...
+}
+
+// change callbacks1!
+function unionOfCallbacks(callbacks1: Callbacks, callbacks2: Callbacks) {
+  for (const callback of callbacks2) {
+    callbacks1.add(callback);
+  }
+}
+
+function createToNotify(): ToNotify {
```
Comment (ged-odoo) [inline src/reactivity.ts:251]: subscriber

File: src/reactivity.ts [hunk 8]
```diff
@@ -143,8 +186,89 @@ export function useState(state: any): Atom {
...
+  return { nodes: new Set(), callbacks: new Set() };
+}
+
+let toNotify: ToNotify | null = null;
```
Comment (ged-odoo) [inline src/reactivity.ts:255]: `let toNotify: Subscribers | null = null

File: src/component/component_node.ts [hunk 9]
```diff
@@ -72,6 +72,7 @@ export class ComponentNode<T extends typeof Component = any> implements VNode<Co
   status: STATUS = STATUS.NEW;
 
   renderFn: Function;
+  parent: ComponentNode | null;
```
Comment (ged-odoo) [inline src/component/component_node.ts:75]: level

File: src/reactivity.ts [hunk 10]
```diff
@@ -102,49 +102,128 @@ function isTrackable(value: any): boolean {
...
+      atom[OBSERVER]();
+    }
   }
-  return unregisterObserver.bind(null, observer);
 }
 
-function registerSource(source: Source) {
-  if (!sourceAtoms.get(source)) {
-    sourceAtoms.set(source, new Set());
+// Need to optimize this!
```
Comment (aab-odoo) [inline src/reactivity.ts:177]: is this the right time to discuss how to optimize this, or is it premature?

File: src/reactivity.ts [hunk 11]
```diff
@@ -102,49 +102,128 @@ function isTrackable(value: any): boolean {
...
+  let current: any = node1;
+  if (current.level <= node2.level) {
+    return false;
   }
+  do {
+    if (current === node2) {
+      return true;
+    }
+    current = current.parent;
+  } while (current);
```
Comment (aab-odoo) [inline src/reactivity.ts:188]: with the current strategy, you could trim the search with
```js
while (current && current.level > node2.level)
```
but maybe it's more costly to always do this check than sometimes go up to the root element

**Review by aab-odoo — COMMENTED** [review]
Do we have an idea of how it behaves perf-wize? It would be nice to know before deploying it.

File: tests/components/concurrency.test.ts [hunk 12]
```diff
@@ -2069,9 +2075,12 @@ test("concurrent renderings scenario 16", async () => {
     "B:render",
     "C:willUpdateProps",
     "C:render",
+    "D:setup",
```
Comment (Polymorphe57) [inline tests/components/concurrency.test.ts:2078]: you need to check this @ged-odoo

**Review by ged-odoo — APPROVED** [review]

File: src/reactivity.ts [hunk 13]
```diff
@@ -102,49 +102,128 @@ function isTrackable(value: any): boolean {
...
   }
-  return unregisterObserver.bind(null, observer);
 }
 
-function registerSource(source: Source) {
-  if (!sourceAtoms.get(source)) {
-    sourceAtoms.set(source, new Set());
+// Need to optimize this!
+function isLessOrEqual(node1: ComponentNode, node2: ComponentNode) {
+  let current: any = node1;
```
Comment (ged-odoo) [inline src/reactivity.ts:179]: let level2 = node2.level

File: src/reactivity.ts [hunk 14]
```diff
@@ -102,49 +102,128 @@ function isTrackable(value: any): boolean {
...
+  if (current.level <= node2.level) {
```
Comment (ged-odoo) [inline src/reactivity.ts:180]: replace node2.level with level2

**Review by ged-odoo — APPROVED** [review]

**Pull request merged**